src/server/tiering/op_manager.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/op_manager.h"
+
+#include "base/logging.h"
+#include "core/overloaded.h"
+#include "io/io.h"
+#include "server/tiering/common.h"
+
+namespace dfly::tiering {
+
+namespace {
+
+OpManager::OwnedEntryId ToOwned(OpManager::EntryId id) {
+  Overloaded convert{[](unsigned i) -> OpManager::OwnedEntryId { return i; },
+                     [](std::string_view s) -> OpManager::OwnedEntryId { return std::string{s}; }};
+  return std::visit(convert, id);
+}
+
+OpManager::EntryId Borrowed(const OpManager::OwnedEntryId& id) {
+  return std::visit([](const auto& v) -> OpManager::EntryId { return v; }, id);
+}
+
+}  // namespace
+
+std::error_code OpManager::Open(std::string_view file) {
+  return storage_.Open(file);
+}
+
+void OpManager::Close() {
+  storage_.Close();
+}
+
+util::fb2::Future<std::string> OpManager::Read(EntryId id, DiskSegment segment) {
+  return PrepareRead(segment).ForId(id, segment).futures.emplace_back().get_future();
+}
+
+void OpManager::Delete(EntryId id) {
+  // If the item isn't offloaded, it has io pending, so cancel it
+  DCHECK(pending_stash_ver_.count(ToOwned(id)));
+  ++pending_stash_ver_[ToOwned(id)];
+}
+
+void OpManager::Delete(DiskSegment segment) {
+  if (auto it = pending_reads_.find(segment.offset); it != pending_reads_.end()) {
+    // If a read is pending, it will be deleted once the read finished
+    DCHECK_EQ(it->second.segment.length, segment.length);
+    it->second.delete_requested = true;
+  } else {
+    // Otherwise, delete it immediately
+    storage_.MarkAsFree(segment);
+  }
+}
+
+std::error_code OpManager::Stash(EntryId id_ref, std::string_view value) {
+  auto id = ToOwned(id_ref);
+  unsigned version = ++pending_stash_ver_[id];
+
+  std::shared_ptr<char[]> buf(new char[value.length()]);
+  memcpy(buf.get(), value.data(), value.length());
+
+  io::MutableBytes buf_view{reinterpret_cast<uint8_t*>(buf.get()), value.length()};
+  return storage_.Stash(buf_view, [this, version, id = std::move(id), buf](DiskSegment segment) {
+    ProcessStashed(Borrowed(id), version, segment);
+  });
+}
+
+OpManager::ReadOp& OpManager::PrepareRead(DiskSegment segment) {
+  // If a read for a small key is requested, read the whole page.
+  // Small keys never begin at 0 because their keys are stored before values.
+  if (segment.offset % 4_KB != 0) {
+    DCHECK_LT(segment.length, 4_KB);
+    segment = {segment.offset / 4_KB * 4_KB, 4_KB};
```

Comment:
you loose the value offset, so how do you read the value from the page?
seems like a bug that requires  a test

Reply:
In Read() I pass it twice

```
PrepareRead(segment).ForId(id, segment)
```

and the first modified segment isn't stored anywhere

ProcessRead will calculate it correctly

Reply:
wow, it's confusing by reading the code in the function.
suggestion: call the argument `aligned_segment`, dcheck on it and pass the aligned segment in `Read`.

Another thing - it will fail for large lengths like 4095 because you never align length. Unless you also handle it somewhere 😆